packages/foam-vscode/src/features/wikilink-provider.ts

```diff
@@ -0,0 +1,183 @@
+import { createNoteForPlaceholderWikilink } from './create-from-template';
+import { OPEN_COMMAND } from './utility-commands';
+import { monitorFoamVsCodeConfig } from '../services/config';
+
+const GO_TO_DEFINITION_NAV_MODE = 'goToDefinition';
+const OPEN_LINK_NAV_MODE = 'openLink';
+const MIXED_NAV_MODE = 'goToDefinitionButOpenLinkForCreation';
+const NAVIGATION_DISABLED = 'off';
+
+const feature: FoamFeature = {
+  activate: async (
+    context: vscode.ExtensionContext,
+    foamPromise: Promise<Foam>
+  ) => {
+    const foam = await foamPromise;
+
+    const wikilinkProvider = new WikilinkProvider(
+      foam.workspace,
+      foam.services.parser,
+      monitorFoamVsCodeConfig('links.navigationMode')
+    );
+
+    context.subscriptions.push(
+      vscode.languages.registerDefinitionProvider(
+        mdDocSelector,
+        wikilinkProvider
+      ),
+      vscode.languages.registerDocumentLinkProvider(
+        mdDocSelector,
+        wikilinkProvider
+      )
+    );
+  },
+};
+
+export class WikilinkProvider
+  implements vscode.DefinitionProvider, vscode.DocumentLinkProvider {
+  constructor(
+    private workspace: FoamWorkspace,
+    private parser: ResourceParser,
+    private getNavigationMode: () => string
+  ) {}
+
+  public async provideDefinition(
+    document: vscode.TextDocument,
+    position: vscode.Position,
+    token: vscode.CancellationToken
+  ): Promise<vscode.LocationLink[] | vscode.Definition> {
+    const navMode = this.getNavigationMode();
+    if (navMode === OPEN_LINK_NAV_MODE || navMode === NAVIGATION_DISABLED) {
+      return;
+    }
+
+    const resource = this.parser.parse(document.uri, document.getText());
+    const targetLink: ResourceLink | undefined = resource.links.find(link =>
+      Range.containsPosition(link.range, {
+        line: position.line,
+        character: position.character,
+      })
+    );
+
+    if (!targetLink) {
+      return;
+    }
+
+    const uri = this.workspace.resolveLink(resource, targetLink);
+    if (URI.isPlaceholder(uri)) {
+      if (
+        navMode === MIXED_NAV_MODE ||
+        vscode.workspace.workspaceFolders.length < 1
+      ) {
+        return;
+      }
+
+      const newNoteUri = toVsCodeUri(
+        URI.createResourceUriFromPlaceholder(
+          vscode.workspace.workspaceFolders[0].uri,
+          uri
+        )
+      );
+      const newNoteTitle = uri.path.split('/').slice(-1)[0];
+      await createNoteForPlaceholderWikilink(newNoteTitle, newNoteUri);
+    }
+
+    const targetResource = this.workspace.get(uri);
+    const targetRange = findFirstLineOfContent(targetResource.source.text);
+
+    const result: vscode.LocationLink = {
+      originSelectionRange: toVsCodeRange(targetLink.range),
+      targetUri: toVsCodeUri(uri),
+      targetRange: null, // Disable the preview feature to avoid duplication with the over feature.
+      targetSelectionRange: toVsCodeRange(targetRange),
```

Comment:
When following a link with the `Go To Definition` command, this allows to briefly highlight the first line of the MD content (usually the title).

Reply:
Is this moving the cursor to that position? what if the note was already open? would it displace the former cursor position?